chore: Fly refactoring; fixes; extra LSN shipping coverage

pcnc opened the pull request:
* Refactors LSN checkpoint creation and handling
  * Introduces handler that regularly ships the checkpoint file - used as a backup for cases such as volume corruption
    * Currently at 10 minute intervals, if the latest LSN has updated in the meantime
  * Ships the checkpoint after supervisord exits, as a backup to when signal handling goes awry
* Some fixes around perms and order of operations when unpacking the init payload

pcnc committed:
chore: fly fixes

dragarcia committed:
chore: bump adminapi and admin-mgr binaries (#792)

* chore: bump admin-api and admin-mgr

* chore: init read-replica.conf
* and ensure it's accounted for everywhere

* chore: bump version

* chore: remove rc versioning

* chore: use admin-api v0.54.0 instead

* chore: include read-replica.conf from start

imor committed:
chore: compile pg_graphql from source for OrioleDb Dockerfile

imor committed:
chore: compile pg_jsonschema from source for OrioleDb Dockerfile

imor committed:
fix: use stable rust toolchain

imor committed:
fix: adjust pg_graphql.deb paths so that the extension loads

imor committed:
fix: adjust pg_jsonschema.deb paths so that the extension loads

darora committed:
chore: add postgres team as owners on orioledb files

tomashley committed:
chore: add some network debugging tools, clean out unused pkgs (#794)

pcnc committed:
chore: configure autoshutdown using remote init payload

pcnc committed:
chore: configure autoshutdown using remote init payload

pcnc committed:
chore: more lsn checkpoint handling

pcnc committed:
chore: more lsn checkpoint handling

bmpandrade committed:
chore: ensure fail2ban stays disabled when FAIL2BAN_DISABLED is set  (#796)

* chore: ensure fail2ban stays disabled when FAIL2BAN_DISABLED is set

* chore: update docker aio entrypoint with extra validation on env vars check

pcnc committed:
chore: more lsn checkpoint handling

pcnc committed:
chore: cleanup

pcnc committed:
chore: address shutdown event listener logic

pcnc committed:
chore: bump pg version

pcnc committed:
Merge branch 'develop' into pcnc/fly-fixes-redux

pcnc committed:
chore: newline

pcnc committed:
chore: newline

darora commented on docker/all-in-one/entrypoint.sh:
```diff
@@ -51,13 +53,57 @@ function enable_swap {
   fi
 }
 
+function push_lsn_checkpoint_file {
```
Should we consider moving the LSN related complexity a level down into the admin-mgr, where it can also be better tested?

darora replied:
Could also consider moving down most of the entrypoint functionality, including the configure scripts etc. Or perhaps adding an image entrypoint binary in this repo, if admin-mgr (and cross-repo for entrypoint tasks) doesn't make sense

pcnc replied:
Good point, going to move LSN handling fully into adminmgr - better for it to be self-contained rather than handled in a handful of other ways. Would simplify the event listener considerably as well

darora commented on docker/all-in-one/entrypoint.sh:
```diff
@@ -160,6 +206,8 @@ function start_supervisor {
 
   # Start supervisord
   /usr/bin/supervisord -c $SUPERVISOR_CONF
+
```
stray newlines

darora commented on docker/all-in-one/shutdown_event_listener.py:
```diff
@@ -0,0 +1,57 @@
+    sys.stderr.flush()
+
+def main():
+    while 1:
```
`while True`?

darora commented on docker/all-in-one/shutdown_event_listener.py:
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/python3
```
Needs tests

darora commented on docker/all-in-one/shutdown_event_listener.py:
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/python3
+
+# This script is used to make sure the latest LSN checkpoint is persisted remotely
+#  before the container is stopped. It is triggered every 60 seconds, whichever comes first.
```
Perhaps I'm misreading the script, but how does any invocation terminate? Or is it a single instance that's spawned, and gets new events fed in over stdin as they occur?

pcnc replied:
It's an ongoing script which supervisord keeps alive then feeds it events, yes.

darora commented on docker/all-in-one/shutdown_event_listener.py:
```diff
@@ -0,0 +1,57 @@
+
+def main():
+    while 1:
+        write_stdout('READY\n')
```
This output is a contract with supervisord?

darora replied:
do we need additional debugging output to differentiate when the script didn't persist things for w/e reason?

pcnc replied:
>This output is a contract with supervisord?

Yep - `READY` means supervisord can send events
http://supervisord.org/events.html#event-listener-states

End-state of an event completing succesfully is 
```
RESULT <resultLen>
<result>
```

>do we need additional debugging output to differentiate when the script didn't persist things for w/e reason?

Yep. Would definitely be worth having it log to its own log file and log whenever it bails on persisting

darora commented on docker/all-in-one/shutdown_event_listener.py:
```diff
@@ -0,0 +1,57 @@
+                with open(checkpointFile) as f:
+                    currentLSN = f.read()
+                    if currentLSN != '0/0' and currentLSN != previousLSN:
+                        subprocess.run(["/usr/bin/admin-mgr", "lsn-checkpoint-push"])
```
I don't think this captures stdout/stderr, or even error out if admin-mgr runs into an error?

pcnc replied:
admin-mgr logs all its output to `/var/log/walg/pitr.log` by default :/

Agreed that it doesn't capture any error states for further debugging. addressing

darora commented:
Testinfra AIO tests were passing in the past I think?

pcnc committed:
Merge remote-tracking branch 'refs/remotes/origin/pcnc/fly-fixes-redux' into pcnc/fly-fixes-redux

pcnc committed:
chore: bump testinfra workflow timeout

bmpandrade committed:
Merge branch 'develop' into pcnc/fly-fixes-redux

bmpandrade commented:
Still failing test-infra-aio with:
```
E           requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url: /rest/v1/buckets (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0xffffbc858f70>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

Looks like services are taking longer to start and it fails. 

@pcnc , will look into the test suite and see if there is a hold we can do until services are up and running, or if the service itself failed to start.

pcnc committed:
chore: print debug logs

pcnc committed:
chore: fix for missing env

pcnc committed:
chore: bump default postgresql_version to 15.0.1.136

hf committed:
feat: bump gotrue to v2.125.1 (#820)

pcnc committed:
chore: additional handling for testinfra AIO container start failure

pcnc commented:
@soedirgo - added additional handling so that the testinfra AIO workflow will attempt checking if the container is alive for a max of 60 times, or until it's marked as `exited`, before outright failing + print container logs: https://github.com/supabase/postgres/blob/pcnc/fly-fixes-redux/testinfra/test_all_in_one.py#L71-L76

This should result in a bit faster feedback timing-wise

pcnc committed:
chore: logging for shutdown event listener

pcnc committed:
chore: fix ci version checker

pcnc committed:
Merge branch 'develop' into pcnc/fly-fixes-redux

pcnc commented:
> Testinfra AIO tests were passing in the past I think?

Fixed - supervisord service entry missing an env var which was added in the later steps of developing the PR.
Also improved the testinfra script to fail whenever the container is marked as `exited` instead of continuously waiting for it to become healthy + output container logs on failure

pcnc committed:
chore: remove incompatible supervisord setting

pcnc committed:
chore: ship logs to file

pcnc committed:
chore: more testinfra output handling

pcnc committed:
chore: testinfra improvements cleanup

pcnc committed:
chore: fix failing container bootup

burmecia committed:
chore: bump wrappers version to 0.2.0

pcnc committed:
chore: testinfra AMI test

pcnc committed:
chore: faster failure for testinfra AMI

pcnc committed:
chore: cleanup resources properly

pcnc committed:
chore: temporarily skip steps for debugging

pcnc committed:
chore: revert skipped steps

pcnc committed:
chore: listener refactoring

pcnc committed:
chore: foo

pcnc committed:
Merge remote-tracking branch 'origin/develop' into pcnc/fly-fixes-redux

pcnc committed:
chore: fix gotrue

pcnc committed:
chore: fix workflow

pcnc committed:
chore: remove supervisord event listener; use built-in admin-mgr watcher

bmpandrade commented on .github/workflows/dockerhub-release.yml:
```diff
@@ -7,7 +7,7 @@ on:
       - ".github/workflows/dockerhub-release.yml"
       - "common.vars*"
-
+      
```
```suggestion
 
```

bmpandrade commented on docker/all-in-one/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-ARG postgres_version=15.1.0.83
+ARG postgres_version=15.1.0.136
```
```suggestion
ARG postgres_version=15.1.0.148
```

Maybe we could bump into latest? Anyway, not a blocker since everything runs again.

pcnc replied:
Good point - better to have an up-to-date default base

bmpandrade commented on docker/all-in-one/entrypoint.sh:
```diff
@@ -50,14 +54,40 @@ function enable_swap {
+}
+
+function enable_lsn_checkpoint_push {
+    sed -i "s/autostart=.*/autostart=true/" /etc/supervisor/db-only/lsn-checkpoint-push.conf
```
Isn't this required even if not on a box configured as `db-only`? 🤔  (at least for all Fly boxes)

Similar thought for the remaining `supa-shutdown` and `lsn-checkpoint-push` files for supervisord or references, that bring up `db-only`.

pcnc replied:
Ah, the `db-only` folder is the default (and minimal) set of services which supervisord runs:
https://github.com/supabase/postgres/blob/develop/docker/all-in-one/etc/supervisor/supervisord.conf#L169-L170

During regular operation if `POSTGRES_ONLY` is not true, `services/*conf` also gets added:
https://github.com/supabase/postgres/blob/develop/docker/all-in-one/entrypoint.sh#L234-L238

Though, good point - this causes quite a bit of confusion - better to rename the `db-only` folder to `default` or `base-services`

bmpandrade replied:
Got it. Thanks for the points it might have escaped me but, yeah, unless you are very comfortable already with the term `db-only` it might be a bit misleading. Change to `base-services`, since the other folder is already called `services` sounds good to me 👌 

Regarding checks/tests:
- `Testinfra Integration Tests / test-all-in-one ` is failing with
```
FAILED testinfra/test_all_in_one.py::test_postgrest_responds_to_requests - requests.exceptions.ConnectionError:
HTTPConnectionPool(host='localhost', port=8000): Max retries exceeded with url:
 /rest/v1/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection 
object at 0x7f8a8357bf70>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

- `Testinfra Integration Tests / test-ami` is failing with 
```
Build 'amazon-ebssurrogate.source' errored after 11 minutes 52 seconds: 
Script exited with non-zero exit status: 2.Allowed exit codes are: [0]
```

bmpandrade reviewed (commented):
In general it looks good, but I got a bit confused about the `db-only` as referred on the other code comments.

I thought, that only on special cases `db-only` was used, and the fly extensions was not one of them. Wouldn't this make us update the worker too, when deploying to Fly to treat it as a `db-only`, to ensure that LSN checkpoints were being properly uploaded?

pcnc committed:
Update .github/workflows/dockerhub-release.yml

Co-authored-by: Bruno Andrade <1145012+bmpandrade@users.noreply.github.com>

pcnc committed:
Update docker/all-in-one/Dockerfile

Co-authored-by: Bruno Andrade <1145012+bmpandrade@users.noreply.github.com>

pcnc committed:
chore: rename `db-only` to `base-services`

thebengeu committed:
feat: add `systemctl try-restart` commands to adminapi sudoers

thebengeu committed:
feat: add `supervisorctl status` commands and other commands for `services:envoy` to adminapi sudoers

`supervisorctl status` is used as part of the supervisord commands equivalent to
`systemctl try-restart`.

thebengeu committed:
fix: set `autorestart` to `true` for Envoy supervisord conf

It's fine to leave it `true` while toggling `autostart`.

thebengeu committed:
feat: upgrade Admin API to 0.58.0

pcnc committed:
chore: bump adminapi and adminmgr versions

pcnc committed:
Merge remote-tracking branch 'origin' into pcnc/fly-fixes-redux

pcnc committed:
chore: remove event listener

pcnc committed:
chore: trigger rc0 build

sweatybridge committed:
fix: use C collation for compiling extensions with orioledb

sweatybridge committed:
chore: add postgres team as codeowner or version file

sweatybridge committed:
fix: bump pgrx version for wrappers build (#843)

steve-chavez committed:
chore: bump postgrest to v12.0.1 (#844)

pcnc committed:
chore: fix tests

pcnc committed:
chore: update AIO image todo (#845)

pcnc committed:
chore: logrotate, more tests

pcnc committed:
chore: cleanup

pcnc committed:
Merge remote-tracking branch 'origin/develop' into pcnc/fly-fixes-redux
